### DISCUSSION TITLE
Add Meter timestamp to MQTT output as a separate attribute called datetime

### DISCUSSION THREAD

**Discussion by zelbanna** [UtilitechAS/amsreader-firmware#889] [Ideas]
I use the AMS reader as the source of truth for electrical metering. Using a MQTT subscriber I republish the readings to other services (instead of lots of devices) e.g. Tibber Pulse :-) My issue is that I would like to avoid guessing the meter's time reading and use the actual meter datetime instead. The request is hence to add an attribute to MQTT (both topics power and energy) so I can use actual meter datetime (from the OBIS field) 

I am fine with decoding and reencoding the 12 byte string myself if not sending the actual datetime string

**Comment by ArnieO** [comment]
One potential issue with that is that we have seen that meter timestamp is sometimes not reliable, for instance regarding Daylight Saving Time. And some meters do not send a timestamp at all. But we could potentially publish it "as is" so that the user can attempt to use it (empty string if no timestamp from the meter).

@gskjold may want to weigh in here!

**Comment by zelbanna** [comment]
That would be perfect!

The Tibber Pulse didn't seem that responsive anyway, mqtt data normally came in 4 seconds after the meter issued the time stamp whereas the Pow-K+ is consistently below 2 seconds (and I am pretty sure my hourly NTP synced MQTT machine doesn't drift in seconds but rather milliseconds...). So anything would be better than that Pulse device IMHO. As a side note: For that hourly "total" reading I even saw 9 seconds delay on the pulse...

**Comment by gskjold** [comment]
Meter timestamp is included in JSON sent to MQTT in the "rtc" field, or on the topic meter/clock if you use raw mode. But not all meters report their own time. The "t" field in JSON is the package timestamp, this is in some cases populated by the meter, but if not we substitute NTP timestamp into this field.

Edit, search for "rtc" and "t" here: https://github.com/UtilitechAS/amsreader-firmware/wiki/Message-formats#list-1

**Reply by zelbanna** [reply]
I am not using raw mode but Home Assistant mode so HA get this (and all other MQTT data from other sources) directly from MQTT server.

I am monitoring the reader topics and reencode power and energy into Tibber Pulse model (List 2 and 3 - meter is setup with HAN RJ45 to supply power for my older Tibber Pulse) and publish into AWS. 

The RTC field does not exist in the power topic message as you say (and rtc field is 0 in the energy message) hence my ask to add meter rtc time stamp also for HA configured MQTT mode (if it exists). That way I wouldn't need to guess from localtime but could use actual time of the issued message

**Reply by gskjold** [reply]
Empty rtc field is probably due to this issue: https://github.com/UtilitechAS/amsreader-firmware/issues/893
I will also consider if we should add meter timestamp and package timestamp to HA MQTT payloads
